File: matlab/ch07/sinevap.m
function res = sinevap(evdat)
% Calculation of single-effect evaporator
% Data
Tc = 647.096; % critical temperature of H2O (K)
Pc = 22064000; % critical pressure of H2O (Pa)
T0 = 273.15; xf = evdat.xf; xp = evdat.xp; mf = evdat.mf;
Tf = (evdat.Tf-32)/1.8 + 273.15; % F->K
Ps = evdat.Ps*6894.757; Pv = evdat.Pv*6894.757; % psia->Pa
U = evdat.U; cf = 1/2326; % J/kg->Btu/lb
% Saturation temperature
a = [-7.85951783 1.84408259 -11.7866497 22.6807411 -15.9618719 1.80122502];
gs = @(x) a(1)*x + a(2)*x^1.5 + a(3)*x^3 + a(4)*x^3.5 + a(5)*x^4 +a(6)*x^7.5 - (1-x)*log(Ps/Pc);
gv = @(x) a(1)*x + a(2)*x^1.5 + a(3)*x^3 + a(4)*x^3.5 + a(5)*x^4 +a(6)*x^7.5 - (1-x)*log(Pv/Pc);
xs = fzero(gs,0.5); xv = fzero(gv,0.5); Ts = Tc*(1 - xs); Tv = Tc*(1 - xv);
% Enthalpy
sts = satsteam(Ts-T0); stv = satsteam(Tv-T0); stf = satsteam(Tf-T0);
Hs = sts.hV*cf; hs = sts.hL*cf; % steam
Hv = stv.hV*cf; hv = stv.hL*cf; % solution
hf = stf.hL*cf;
% Material balance
mp = xf*mf/xp; % Btu/hr
mev = mf - mp; q = mev*Hv - mf*hf + mp*hv; ms = q/(Hs - hs);
Ts = (Ts - 273.15)*1.8+32; % K->F
Tv = (Tv - 273.15)*1.8+32; dT = Ts - Tv; A = q/U/dT;
% Assign output structure (heating surface(A) and steam amount(ms))
res.ms = ms; res.A = A;
end

File: python/ch07/sinevap.py
import numpy as np
from scipy.optimize import fsolve
from satsteam import satsteam

def sinevap(evdat):
    """
    단일 효과 증발기(Single-effect evaporator) 계산 함수
    """
    # 상수 및 데이터 준비
    Tc = 647.096  # K
    Pc = 22064000  # Pa
    T0 = 273.15
    xf = evdat['xf']
    xp = evdat['xp']
    mf = evdat['mf']
    Tf = (evdat['Tf'] - 32) / 1.8 + T0
    Ps = evdat['Ps'] * 6894.757
    Pv = evdat['Pv'] * 6894.757
    U = evdat['U']
    cf = 1 / 2326  # J/kg -> Btu/lb
    
    # Saturation temperature 계산을 위한 다항식 계수
    a = np.array([-7.85951783, 1.84408259, -11.7866497, 22.6807411, -15.9618719, 1.80122502])
    
    def get_x(P):
        # f(x) = 0을 만족하는 x를 찾기 위한 함수
        func = lambda x: (a[0]*x + a[1]*x**1.5 + a[2]*x**3 + a[3]*x**3.5 + 
                          a[4]*x**4 + a[5]*x**7.5 - (1-x)*np.log(P/Pc))
        return fsolve(func, 0.5)[0]
    
    xs = get_x(Ps)
    xv = get_x(Pv)
    Ts_k = Tc * (1 - xs)
    Tv_k = Tc * (1 - xv)
    
    # 엔탈피 계산 (외부 satsteam 함수 호출)
    sts = satsteam(Ts_k - T0)
    stv = satsteam(Tv_k - T0)
    stf = satsteam(Tf - T0)
    
    Hs = sts['hV'] * cf
    hs = sts['hL'] * cf
    Hv = stv['hV'] * cf
    hv = stv['hL'] * cf
    hf = stf['hL'] * cf
    
    # 물질 수지 및 열 수지 계산
    mp = xf * mf / xp
    mev = mf - mp
    q = mev * Hv - mf * hf + mp * hv
    ms = q / (Hs - hs)
    
    # 결과 계산 (단위 변환)
    Ts_f = (Ts_k - T0) * 1.8 + 32
    Tv_f = (Tv_k - T0) * 1.8 + 32
    dT = Ts_f - Tv_f
    A = q / U / dT
    
    # 결과 반환
    res = {
        'ms': ms,
        'A': A
    }
    return res